### PR TITLE
8263577: C2: reachable nodes shouldn't have dead uses at the end of optimizations

### PR DESCRIPTION
**Pull request by rwestrel**
It's not uncommon for c2 to leave dead uses hanging from reachable
nodes. It's usually harmless but can cause confusion (such as in
valhalla JDK-8262289, not reproducible with non valhalla code AFAICT).

Most of the time, these are caused by calls to set_req() during igvn
that should be set_req_X() calls. I added a verification pass at the
end of optimization and fixed all the ones I found.

/cc hotspot-compiler

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263577](https://bugs.openjdk.java.net/browse/JDK-8263577): C2: reachable nodes shouldn't have dead uses at the end of optimizations

### Reviewers
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to f214361e3114f3471c9e47c31216efd318669a31
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**) ⚠️ Review applies to f214361e3114f3471c9e47c31216efd318669a31

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/3012/head:pull/3012`
`$ git checkout pull/3012`

### CODE REVIEW

File: src/hotspot/share/opto/node.hpp [hunk 1]
```diff
@@ -454,8 +454,8 @@ class Node {
     }
     return -1;
   }
-  int replace_edge(Node* old, Node* neww);
-  int replace_edges_in_range(Node* old, Node* neww, int start, int end);
+  int replace_edge(Node* old, Node* neww, PhaseGVN* gvn = NULL);
+  int replace_edges_in_range(Node* old, Node* neww, int start, int end, PhaseGVN* gv);
```
Comment (iwanowww) [inline src/hotspot/share/opto/node.hpp:458]: s/gv/gvn/

**Review by iwanowww — APPROVED** [review]
Looks good.

Changes in `src/hotspot/share/opto/subtypenode.cpp` don't look relevant. I assume new verification logic uncovered a problem there. Does it make sense to integrate the fix separately?

**Comment by rwestrel** [comment]
> Changes in `src/hotspot/share/opto/subtypenode.cpp` don't look relevant. I assume new verification logic uncovered a problem there. Does it make sense to integrate the fix separately?

In the case of subtypenode.cpp, some nodes were created and never attached to any use.

File: src/hotspot/share/opto/node.hpp (hunk 1 above)
Comment (rwestrel) [reply src/hotspot/share/opto/node.hpp:458]: right. Will fix that.

**Comment by iwanowww** [comment]
> In the case of subtypenode.cpp, some nodes were created and never attached to any use.

Ok, makes sense. 

So, you move the code out of `if (super_t->singleton() && ...) {` to avoid putting `subklass` on the worklist?

**Comment by rwestrel** [comment]
> So, you move the code out of `if (super_t->singleton() && ...) {` to avoid putting `subklass` on the worklist?

Right.

**Review by iwanowww — APPROVED** [review]

File: src/hotspot/share/opto/addnode.cpp [hunk 2]
```diff
@@ -145,14 +145,8 @@ Node *AddNode::Ideal(PhaseGVN *phase, bool can_reshape) {
...
       Node *x2 = phase->makecon(add1->as_Add()->add_ring(t2, t12));
-      PhaseIterGVN *igvn = phase->is_IterGVN();
-      if (igvn) {
-        set_req_X(2,x2,igvn);
-        set_req_X(1,x1,igvn);
-      } else {
-        set_req(2,x2);
-        set_req(1,x1);
-      }
+      set_req_X(2,x2,phase);
```
Comment (neliasso) [inline src/hotspot/share/opto/addnode.cpp:148]: Missing space after comma in argument list. Same a bit further down to.

**Review by neliasso — APPROVED** [review]

File: src/hotspot/share/opto/addnode.cpp (hunk 2 above)
Comment (rwestrel) [reply src/hotspot/share/opto/addnode.cpp]: Thanks for the review. Should be fixed now.

**Comment by navyxliu** [comment]
hi, @rwestrel 

Is Node::set_req_X() same as PhaseIterGVN::replace_input_of()? 
This even supports delayed hash. can I use them interchangeably? 

```
  // Replace ith edge of "n" with "in"
  void replace_input_of(Node* n, int i, Node* in) {
    rehash_node_delayed(n);
    n->set_req_X(i, in, this);
  }
```

**Comment by rwestrel** [comment]
> hi, @rwestrel
> 
> Is Node::set_req_X() same as PhaseIterGVN::replace_input_of()?
> This even supports delayed hash. can I use them interchangeably?

They are not quite the same as replace_input_of() has the rehash_node_delayed() call. When modifying an existing node, it's a requirement to call has_delete() which rehash_node_delayed() does for you. If you're in Ideal() modifying the "this" node, the hash_delete() call happened before the ideal call.

So I guess roughly: in Ideal modifying this's input -> set_req_X(), otherwise replace_input_of()

**Comment by navyxliu** [comment]
@rwestrel 
I see.  Thanks for the explanation.
verify that your patch invokes "set_req_X" in Node::ideal() and its overrides.

**Pull request closed without merging**